scps/SCP-6.md

```diff
@@ -0,0 +1,134 @@
+
+Skipping the version field `v`, which is as barebones as it gets, and moving onto `id`. This field can be set and used by Ipfs-log users to identify and separate entries made to different logs, it is meant to be a string. There is not much to change here as it's just a string set by the user in userland (like in OrbitDB). However in entry version 3 `id` is renamed to `log` because id is used in a couple different places and log seems much better at describing the property.
+
+The `clock` field's value is an object with two properties, `time` and `id`. It is a Lamport clock logical timestamp and is the heaviest ordering metric of entries by default. The `clock.time` field, like `v`, is simply a positive integer, however `clock.id` by default is set to identity.publicKey which already exists in other places in the entry. This is a good chunk of bytes that can be excluded, either by lazily adding `clock.id` when explicitly set by the user or removing it completely (changing `clock` to what `clock.time` is now).
+
+Fields relating to writer authentication are where the bulk of the bytes can be trimmed from the entry. The `key` field contains the writer's publicKey from their identity. This publicKey also exists at `identity.publicKey` so `key` could be removed from the entry.
+
+As mentioned in [The Road to 1.0](https://github.com/orbitdb/orbit-db/issues/819), `separate identities/keys from oplog entry. CID per identity/key. cuts N bytes from each pubsub message and bitswap/ipfs/ild block transfer.`, the `identity` field could be replaced with the CID of the identity. This deduplication should work well as the `identity` field is the same data in every entry for each unique writer. However in situations where writer keys are ephemeral this is less advantaged and may be a good area to leave flexible.
+
+The `sig` field contains a string, hex encoded digital signature of fields: `v`, `clock`, `id`, `payload`, `next`, `refs`, and `hash`. It is used to verify that `identity` is the one that created the entry and that it wrote the entry. Since the data is currently a hex encoded string, it could instead be encoded into base64 strings to pinch a few bytes (~40B per `sig`). Base64 uses [33% less space](https://stackoverflow.com/questions/3183841/base64-vs-hex-for-sending-binary-content-over-the-internet-in-xml-doc) than base16/hex so if encoding/decoding speeds are similar this change may make sense.
+
+Entry version 3 does not propose any changes to fields `payload`, `next`, or `refs`.
+
+### Entry version 3
+
+What's changed?
+
+```
+{
+  v: 3,
+  log: String,
+  clock: Number,
+  writer: {
+    id: <CID>,
+    sig: String
+  },

```

Comment:
How about re-arraging this so that `sig` is "on the root level", and we rename `id` to `auth` (as in "this CID contains the authentication information") and put it on the root level as well. This would flatten the structure further (together with the proposal below to merge next and refs):
```
{
  v: 3,
  log: String,
  clock: Number,
  auth: <CID>,
  sig: String,
  payload: <...>,
  refs: [<CID>]
}
```

Reply:
Another optimization that we could consider to pinch a few more bytes (20 chars/bytes per entry) is to rename all these field to be single character, eg. `log`->`l`, `auth`->`a` and so on. However, not sure if this already happens on the IPLD serialization level. If not, it's worth thinking about.

Reply:
> How about re-arraging this further so that `sig` is "on the root level", and we rename `id` to `auth` (as in "this CID contains the authentication information") and put it on the root level as well. This would flatten the structure further (together with the proposal below to merge next and refs):
> 
> ```
> {
>   v: 3,
>   log: String,
>   clock: Number,
>   auth: <CID>,
>   sig: String,
>   payload: <...>,
>   refs: [<CID>]
> }
> ```

This looks good I will make the updates. I'm fine with moving sig back to the top level as I can't think of a situation where digital signatures wouldnt be used here. a flat structure is also more appealing and comes with less bytes added.


> Another optimization that we could consider to pinch a few more bytes (20 chars/bytes per entry) is to rename all these field to be single character, eg. `log`->`l`, `auth`->`a` and so on. However, not sure if this already happens on the IPLD serialization level. If not, it's worth thinking about.

this is the cbor implementation `ipld-dag-cbor` uses: [borc](https://github.com/dignifiedquire/borc). looking at the source code object keys in CBOR are encoded in utf-8 if they are strings when [writing the key to the buffer](https://github.com/dignifiedquire/borc/blob/f8b6b7630250e2ee5509f7104529490a388c2cc7/src/encoder.js#L482) so this change seems like it would have effect. there is the issue of object keys becoming ambiguous at this level, however at higher levels that can be remedied easily by mapping shortened keys to more descriptive ones. I'm neutral on this change so let me know if you think i should definitely add it!


A change I thought about but never added had to do with the `log` field. Since ipfs-log is casual i would love to explore in what situations the `log` field is absolutely necessary. It may be necessary in every entry like it is used now; however since log entries reference hashes of previous entries only log tails may need to set this field. It's likely this doesnt provide enough protection against replay or its too difficult to verify (needing to traverse all the way back to a log tail). In that case this field should always be set to the address.root/manifest CID for some nice encoding benefits thanks to the [cbor tag field 42](https://www.iana.org/assignments/cbor-tags/cbor-tags.xhtml) reserved for CIDs. This is loosely related to your suggested change in [The Road to 1.0](https://github.com/orbitdb/orbit-db/issues/819) when talking about addresses being stripped down to just CIDs.

Reply:
> there is the issue of object keys becoming ambiguous at this level, however at higher levels that can be remedied easily by mapping shortened keys to more descriptive ones.

agreed, and I like the prposal here (eg. map the keys in the data to "nicer" keys in the in-memory data structure/object).

I'm also not fairly neutral on that, so perhaps we can leave it as is for now, and make decision once we implement it or even leave for later optimizations (post 1.0). what do you think?

Reply:
wrt. to the `log` I see what you mean. I feel we would be better off keeping it consistent, ie. keep it as it is, so that all entries have the log/id. it's the canonical "which log does an entry belong to" and having it there has the benefit of being able to verify each entry stand-alone (without having to do the traversal as you explain in your comment).

Reply:
> I'm also not fairly neutral on that, so perhaps we can leave it as is for now, and make decision once we implement it or even leave for later optimizations (post 1.0). what do you think?

Im fine with leaving this for now as well.

> wrt. to the `log` I see what you mean. I feel we would be better off keeping it consistent, ie. keep it as it is, so that all entries have the log/id. it's the canonical "which log does an entry belong to" and having it there has the benefit of being able to verify each entry stand-alone (without having to do the traversal as you explain in your comment).

what do you think about changing the `log` field from `orbitdb/<CID>/<dbname>` to just `<CID>`